src/Farmer/Builders/Builders.AzureFirewall.fs

```diff
@@ -0,0 +1,97 @@
+                | _ -> ()
+                match this.VirtualHub with
+                | Some (Managed resId) -> resId
+                | _ -> ()
+              ] |> Set.ofList
+              FirewallPolicy = this.FirewallPolicy |> Option.map (fun x -> match x with | Managed resId | Unmanaged resId -> resId)
+              VirtualHub = this.VirtualHub |> Option.map (fun x -> match x with | Managed resId | Unmanaged resId -> resId)
+              HubIPAddresses = this.HubIPAddressSpace |> Option.map (fun x -> x.Arm)
+              Sku = this.Sku }
+            ]
+
+type AzureFirewallBuilder() =
+    member _.Yield _ =
+        { Name = ResourceName.Empty
+          Sku = {Name = SkuName.AZFW_Hub; Tier = SkuTier.Standard}
+          FirewallPolicy = None
+          VirtualHub = None
+          HubIPAddressSpace = None }
+    /// The name of the firewall.
+    [<CustomOperation "name">] member _.Name(state:AzureFirewallConfig, name) = { state with Name = ResourceName name }
+    /// The SKU of the firewall.
+    [<CustomOperation "sku">]
+    member _.Sku(state:AzureFirewallConfig, name, tier) = { state with Sku = {Name = name; Tier = tier} }
+
+    /// Configure this firewall to use an unmanaged firewall policy.
+    [<CustomOperation "link_to_unmanaged_firewall_policy">]
+    member this.LinkToUnmanagedFirewallPolicy (state:AzureFirewallConfig, firewallPolicyName:ResourceName) =
+        { state with FirewallPolicy = Some (Unmanaged(azureFirewallPolicies.resourceId firewallPolicyName)) }
+    member this.LinkToUnmanagedFirewallPolicy (state:AzureFirewallConfig, firewallPolicyName) =
+        { state with FirewallPolicy = Some (Unmanaged(azureFirewallPolicies.resourceId(ResourceName firewallPolicyName))) }
+    /// Configure this firewall to use a managed firewall policy.
+    [<CustomOperation "link_to_firewall_policy">]
+    member this.LinkToFirewallPolicy (state:AzureFirewallConfig, firewallPolicyName:ResourceName) =
+        { state with FirewallPolicy = Some (Managed(azureFirewallPolicies.resourceId firewallPolicyName)) }
+    member this.LinkToFirewallPolicy (state:AzureFirewallConfig, firewallPolicyName) =
+        { state with FirewallPolicy = Some (Managed(azureFirewallPolicies.resourceId(ResourceName firewallPolicyName))) }
+    /// The unmanaged virtualHub to which the firewall belongs. 
+    [<CustomOperation "link_to_unmanaged_vhub">]
+    member this.LinkToUnmanagedVirtualHub (state:AzureFirewallConfig, vhubName:ResourceName) =
+        { state with VirtualHub = Some (Unmanaged(virtualHubs.resourceId vhubName)) }
+    member this.LinkToUnmanagedVirtualHub (state:AzureFirewallConfig, vhubName) =
+        { state with VirtualHub = Some (Unmanaged(virtualHubs.resourceId(ResourceName vhubName))) }
+    /// The managed virtualHub to which the firewall belongs
+    [<CustomOperation "link_to_vhub">]
+    member this.LinkToVirtualHub (state:AzureFirewallConfig, vhubName:ResourceName) =
```

Comment:
If this is a managed vhub, rather than a resource name, take in the Builder / IArmResource itself.

Reply:
Done